service/history/replication/task_store.go

```diff
@@ -0,0 +1,223 @@
+			backoff.WithRetryPolicy(retryPolicy),
+			backoff.WithRetryableError(persistence.IsTransientError),
+		),
+		scope:       metricsClient.Scope(metrics.ReplicatorCacheManagerScope),
+		logger:      logger.WithTags(tag.ComponentReplicationCacheManager),
+		rateLimiter: quotas.NewDynamicRateLimiter(config.ReplicationTaskGenerationQPS.AsFloat64()),
+	}
+}
+
+// Get will return a hydrated replication message for a given cluster based on raw task info.
+// It will either return it immediately from cache or hydrate it, store in cache and then return.
+//
+// Returned task may be nil. This may be due domain not existing in a given cluster or replication message is not longer relevant.
+// Either case is valid and such replication message should be ignored and not returned in the response.
+func (m *TaskStore) Get(ctx context.Context, cluster string, info persistence.ReplicationTaskInfo) (*types.ReplicationTask, error) {
+	cache, ok := m.clusters[cluster]
+	if !ok {
+		return nil, errUnknownCluster
+	}
+
+	domain, err := m.domains.GetDomainByID(info.DomainID)
+	if err != nil {
+		return nil, fmt.Errorf("resolving domain: %w", err)
+	}
+
+	// Domain does not exist in this cluster, do not replicate the task
+	if !domain.HasReplicationCluster(cluster) {
+		return nil, nil
+	}
+
+	scope := m.scope.Tagged(metrics.SourceClusterTag(cluster))
+
+	scope.IncCounter(metrics.CacheRequests)
+	sw := scope.StartTimer(metrics.CacheLatency)
+	defer sw.Stop()
+
+	task := cache.Get(info.TaskID)
+	if task != nil {
+		scope.IncCounter(metrics.CacheHitCounter)
+		return task, nil
+	}
+
+	m.scope.IncCounter(metrics.CacheMissCounter)
+
+	// Rate limit to not kill the database
+	m.rateLimiter.Wait(ctx)
+
+	err = m.throttleRetry.Do(ctx, func() error {
+		var err error
+		task, err = m.hydrator.Hydrate(ctx, info)
+		return err
+	})
+
+	if err != nil {
+		m.scope.IncCounter(metrics.CacheFailures)
+		return nil, err
+	}
+
+	m.Put(task)
+
+	return task, nil
+}
+
+// Put will try to store hydrated replication to all cluster caches.
+// Tasks may not be relevant, as domain is not enabled in some clusters. Ignore task for that cluster.
+// Some clusters may be already have full cache. Ignore task, it will be fetched and hydrated again later.
+// Some clusters may already acknowledged such task. Ignore task, it is no longer relevant for such cluster.
+func (m *TaskStore) Put(task *types.ReplicationTask) {
+	// Do not store nil tasks
+	if task == nil {
+		return
+	}
+
+	domain, err := m.getDomain(task)
+	if err != nil {
+		m.logger.Error("failed to resolve domain", tag.Error(err))
+		return
+	}
+
+	for cluster, cache := range m.clusters {
+		if domain != nil && !domain.HasReplicationCluster(cluster) {
+			continue
+		}
+
+		scope := m.scope.Tagged(metrics.SourceClusterTag(cluster))
+
+		err := cache.Put(task)
+		switch err {
+		case errCacheFull:
+			scope.IncCounter(metrics.CacheFullCounter)
+		case errAlreadyAcked:
+			// No action, this is expected.
```

Comment:
Is this branch needed? How could a new history event be acked by a remote cluster?

Reply:
It could be removed, but I kept it to explicitly cover all possible errors returned by the cache.

Suppose we have current cluster X and two polling clusters A and B.  Both have cache size 3. Reads are in batch size 2.
1. X commit tasks 1, 2.
2. A reads until 2, hydrates them and puts to cache A and cache B. size(A) = 2, size(B) = 2
3. A acks until 2, size(A) = 0, ack(A) = 2
4. X commit task 3, 4.
5. A reads until 4, hydrates them and puts to cache A and cache B. size(A) = 2, size(B) = 3. Task4 is skipped with errCacheFull in cache B.
6. A acks until 4. size(A) = 0, ack(A) = 4
7. B reads until 2, returns from cache. size(B) = 3
8. B acks until 2, size(B) = 1, ack(B) = 2
9. B reads until 4, returns Task3 from cache, hydrates Task4 and puts to cache A and cache B. Cache A skips it with errAlreadyAcked, since ack(A) = 4 > Task3. size(B) = 2
10. B acks until 4, size(B) = 0, ack(B) = 4

To sum up. Some polling cluster may be too slow and gets it cache full. After hydrating caches for faster polling clusters will reject them, since they were already acked by them. Let me know if you need further clarification.